framework/auth/views.py

```diff
@@ -199,87 +199,109 @@ def forgot_password_post(auth, **kwargs):
+    data = {}
+    campaign = request.args.get('campaign')
+    service_url = request.args.get('next')
+    log_out = request.args.get('logout')
+    must_login_warning = False
 
-        GET '/login/?campaign=instituion:
-            if user is logged in, redirect to 'dashboard'
-            show institution login
+    # If `?next=[service_url]` is not provided, redirect to `/dashboard`
+    if not service_url:
+        service_url = web_url_for('dashboard', _absolute=True)
 
-        GET '/login/?campaign=prereg:
-            if user is logged in, redirect to prereg home page
-            else show sign up page and notify user to sign in, set next to prereg home page
+    # Only allow redirects which are relative root or full domain, disallows external redirects.
+    if not (service_url[0] == '/' or
+            service_url.startswith(settings.DOMAIN) or
+            service_url.startswith(settings.CAS_SERVER_URL) or
+            service_url.startswith(settings.MFR_SERVER_URL)):
+        raise HTTPError(http.InvalidURL)
 
-        GET '/login/?next=next_url:
-            if user is logged in, redirect to next_url
-            else redirect to CAS login page with next_url as target service
-    """
+    # Campaign
+    if campaign:
+        if campaign in campaigns.CAMPAIGNS:
+            # `GET /register?campaign=prereg`,
+            # Overwrite CAS service url with dedicated campaign url and set `must_login_warn` notification
+            if campaign != 'institution':
+                service_url = campaigns.campaign_url_for(campaign)
+                data['campaign'] = campaign
+                must_login_warning = True
+            # `GET /register?campaign=institution`,
+            # Redirect to CAS institution login with service url as target service
+            else:
+                return redirect(cas.get_login_url(service_url, campaign='institution'))
+        # raise `http.NOT_FOUND` if campaign is invalid
+        else:
+            return HTTPError(http.NOT_FOUND)
 
-    campaign = request.args.get('campaign')
-    next_url = request.args.get('next')
-    log_out = request.args.get('logout')
-    must_login_warning = True
+    # If user is already logged in, redirect to service url
+    if not log_out and auth.logged_in:
+        return redirect(service_url)
 
-    if not campaign and not next_url and not log_out:
+    # Handle different auth situations for `website.project.views.claim_user_register()`
+    if log_out:
         if auth.logged_in:
-            return redirect(web_url_for('dashboard'))
-        return redirect(cas.get_login_url(web_url_for('dashboard', _absolute=True)))
+            return auth_logout(redirect_url=request.url)
+        else:
+            must_login_warning = True
 
-    if campaign:
-        next_url = campaigns.campaign_url_for(campaign)
+    if must_login_warning:
+        status.push_status_message(language.MUST_LOGIN, trust=False)
```

Comment:
As we think about switching to a SPA for the front end, we might want to think about separating out the login logic from the view logic. Consider re-writing this view so that everything that's not dependent on this going into a mako template is in one function that this view calls, and have that function return a dictionary with the important information (such as which page it's being redirected to or whatever). That will also let you unit test the login functionality easily without having to go through the views, which will potentially make those tests better and faster if you don't use an OSFTestCase.